redisvl/schema/schema.py

```diff
@@ -0,0 +1,306 @@
+import os
+import yaml
+
+from enum import Enum
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Union
+
+from pydantic import BaseModel, ValidationError
+
+from redisvl.schema.fields import (
+    TagFieldSchema,
+    TextFieldSchema,
+    NumericFieldSchema,
+    FlatVectorFieldSchema,
+    HNSWVectorFieldSchema,
+    GeoFieldSchema
+)
+
+from redisvl.utils.utils import (
+    convert_bytes,
+    make_dict,
+)
+
+
+class StorageType(Enum):
+    HASH = "hash"
+    JSON = "json"
+
+
+class IndexModel(BaseModel):
+    """Represents the schema for an index, including its name, optional prefix,
+    and the storage type used."""
+    name: str
+    prefix: str = "rvl"
+    key_separator: str = ":"
+    storage_type: StorageType = StorageType.HASH
+
+class FieldsModel(BaseModel):
+    tag: Optional[List[TagFieldSchema]] = None
+    text: Optional[List[TextFieldSchema]] = None
+    numeric: Optional[List[NumericFieldSchema]] = None
+    geo: Optional[List[GeoFieldSchema]] = None
+    vector: Optional[List[Union[FlatVectorFieldSchema, HNSWVectorFieldSchema]]] = None
+
+
+class SchemaValidationError(Exception):
```

Comment:
ok so the thoughts on validation
- we may need to validate on more than just init. i.e. after a new vector field is added. 
- We should ideally be relying on the pydantic validators here and throwing our exception if we choose to make it cleaner.
- starting our own line of exceptions takes on a whole new part of this library that wasn't in the original design. Are we sure we don't just want to throw ``ValueError``? If we do take on exceptions, I think it should be in a separate PR where we introduce docs and use them where appropriate.

Reply:
agreed. will move to ValueError for now.

we should maybe take on exceptions later, and agreed in a different PR.

so if we remove the `IndexModel` and `FieldsModel` then we don't have validators anymore